### ISSUE TITLE
Take enum values into account when deriving structs?

### ISSUE THREAD

**Issue by RReverser** [serde-rs/serde#1142]
When I have a unit struct like

```rust
const LESS: u8 = 1;
const GREATER: u8 = 2;
const EQUAL: u8 = 4;

#[repr(u8)]
#[derive(Serialize, Deserialize)]
pub enum OrderingOp {
    Equal = EQUAL,
    NotEqual = LESS | GREATER,
    GreaterThanEqual = GREATER | EQUAL,
    LessThanEqual = LESS | EQUAL,
    GreaterThan = GREATER,
    LessThan = LESS,
}
```

Would it make sense to change derived implementation to use actual values for the binary encoding? It would not only make encoding/decoding cheaper, but also give better stability to the binary representation for such enums, where reordering or inserting new enum items can currently shift all the indices giving a different meaning to the same binary items.

I understand that having stable schemas is not a concern of serde in general, but thought it could be still helpful if it would provide such basics.

**Comment by Marwes** [comment]
This is (more or less) https://github.com/serde-rs/serde/issues/497 .

**Comment by oli-obk** [comment]
Jup. There's no difference between literals, constants and constant computations

**Comment by RReverser** [comment]
@oli-obk One difference is that, as far as I understand, it would call `serialize_[type]` instead of `serialize_unit_variant`, while I would like to keep using variant name for human-readable serializers/deserializers, and integer value for binary ones. Would it be possible with that proposal?

**Comment by RReverser** [comment]
TL;DR - all that I want is `variant_index` to be replaced with a specified value, just like it is in normal Rust coercions (order-based index is not very useful on its own when variant has own specified value).

**Comment by oli-obk** [comment]
the advantage of order-based indexing is that the enum can be packed in the fewest bits necessary to encode the difference. So there is a use case.

Wrt the human-readable output, it makes sense, I'll add a comment to the other issue.
